### PR TITLE
feat: Validity for Item taxes

### PR DESCRIPTION
**Pull request by deepeshgarg007**
image

Docs: https://github.com/frappe/erpnext_com/pull/637

### CODE REVIEW

File: erpnext/stock/get_item_details.py [hunk 1]
```diff
@@ -396,6 +405,20 @@ def get_item_tax_template(args, item, out):
 			item_group = item_group_doc.parent_item_group
 
 def _get_item_tax_template(args, taxes, out):
+	taxes_with_validity = []
+
+	for tax in taxes:
+		if tax.valid_from:
+			# In purchase Invoice first preference will be given to supplier invoice date
+			# if supplier date is not present then posting date
+			if args.get('bill_date') and tax.valid_from >= getdate(args.get('bill_date')):
```
Comment (nabinhait) [inline erpnext/stock/get_item_details.py]: tax.valid_from <= getdate(args.get('bill_date'))

File: erpnext/stock/get_item_details.py [hunk 2]
```diff
@@ -396,6 +405,20 @@ def get_item_tax_template(args, item, out):
...
+				taxes_with_validity.append(tax)
+			elif not args.get('bill_date') and args.get('posting_date') and tax.valid_from >= getdate(args.get('posting_date')):
+				taxes_with_validity.append(tax)
+
+	if taxes_with_validity:
+		taxes = taxes_with_validity
```
Comment (nabinhait) [inline erpnext/stock/get_item_details.py]: `taxes_with_validity` should be ordered by `valid_from` desc and while filtering based on `tax_category` it should take latest one.

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Review by nabinhait — CHANGES_REQUESTED** [review]
- In transactions
    - item_tax_template link field's value should be filtered based on validity
    - Validate in server side

**Comment by deepeshgarg007** [comment]
Working on changes in docs. Will update the PR link in description soon

File: erpnext/public/js/controllers/transaction.js [hunk 3]
```diff
@@ -1693,6 +1693,29 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
...
+	set_query_for_item_tax_template: function(doc, cdt, cdn) {
+
+		var item = frappe.get_doc(cdt, cdn);
+		if(!item.item_code) {
+			frappe.throw(__("Please enter Item Code to get item taxes"));
+		} else {
+
+			let filters = {
+				'item_code': item.item_code,
+				'valid_from': doc.posting_date
```
Comment (nabinhait) [inline erpnext/public/js/controllers/transaction.js]: transaction_date in sales/purchase order

File: erpnext/controllers/queries.py [hunk 4]
```diff
@@ -508,3 +508,16 @@ def get_purchase_invoices(doctype, txt, searchfield, start, page_len, filters):
 		query += " and piitem.item_code = {item_code}".format(item_code = frappe.db.escape(filters.get('item_code')))
 
 	return frappe.db.sql(query, filters)
+
+@frappe.whitelist()
+def get_tax_template(doctype, txt, searchfield, start, page_len, filters):
+
+	item_doc = frappe.get_cached_doc('Item', filters.get('item_code'))
+
+	if not item_doc.taxes:
```
Comment (nabinhait) [inline erpnext/controllers/queries.py]: if no taxes in item, check in item_group. if not there as well return all.

File: erpnext/controllers/taxes_and_totals.py [hunk 5]
```diff
@@ -43,6 +45,28 @@ def _calculate(self):
...
+				item_doc = frappe.get_cached_doc("Item", item.item_code)
+				if item.get('item_tax_template'):
+					args = {
+						'tax_category': self.doc.get('tax_category'),
+						'posting_date': self.doc.get('posting_date'),
+						'bill_date': self.doc.get('bill_date')
+					}
+
+					#Do not validate tax if no template mentioned in Item Tax
+					if not item_doc.taxes:
```
Comment (nabinhait) [inline erpnext/controllers/taxes_and_totals.py]: if taxes not in item, check for item_group

File: erpnext/controllers/taxes_and_totals.py [hunk 6]
```diff
@@ -43,6 +45,28 @@ def _calculate(self):
...
 		self.calculate_total_net_weight()
 
+	def validate_item_tax_template(self):
+		for item in self.doc.get('items'):
+			if item.item_code:
+				item_doc = frappe.get_cached_doc("Item", item.item_code)
+				if item.get('item_tax_template'):
+					args = {
+						'tax_category': self.doc.get('tax_category'),
+						'posting_date': self.doc.get('posting_date'),
```
Comment (nabinhait) [inline erpnext/controllers/taxes_and_totals.py]: posting_date or transaction_date?

**Review by nabinhait — CHANGES_REQUESTED** [review]
- tests

File: erpnext/controllers/queries.py [hunk 7]
```diff
@@ -508,3 +508,29 @@ def get_purchase_invoices(doctype, txt, searchfield, start, page_len, filters):
...
+	else:
+		args = {
+			'item_code': filters.get('item_code'),
+			'posting_date': filters.get('valid_from'),
+			'tax_category': filters.get('tax_category')
+		}
+
+		taxes = _get_item_tax_template(args, taxes, for_validate=True)
+
+		return frappe.db.sql(""" SELECT name FROM `tabItem Tax Template` where name
```
Comment (nabinhait) [inline erpnext/controllers/queries.py]: Why are we doing this query? We already have list of valid tax templates.

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Review by nabinhait — APPROVED** [review]

File: erpnext/public/js/controllers/transaction.js [hunk 8]
```diff
@@ -1700,6 +1700,29 @@ erpnext.TransactionController = erpnext.taxes_and_totals.extend({
 		}
 	},
 
+	set_query_for_item_tax_template: function(doc, cdt, cdn) {
+
+		var item = frappe.get_doc(cdt, cdn);
+		if(!item.item_code) {
+			frappe.throw(__("Please enter Item Code to get item taxes"));
```
Comment (nameduser0) [inline erpnext/public/js/controllers/transaction.js:1707]: Purchase and Sales Invoice lines should never need an item code for anything, to support one-off manually typed lines

**Pull request merged**